server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,539 @@
+				if (!hideLogin) {
+					loginTimes[toID(friendID)] = last_login;
+				}
+			}
+		}
+
+		const sorted = Object.keys(categorized)
+			.filter(item => categorized[item].length > 0)
+			.map(item => `${STATUS_TITLES[item]} (${categorized[item].length})`);
+
+		let buf = `<h3>Your friends: <small> `;
+		if (sorted.length > 0) {
+			buf += `Total (${friends.length}) | ${sorted.join(' | ')}`;
+		} else {
+			buf += `</h3><em>you have no friends added on Showdown lol</em><br /><br /><br />`;
+			buf += `<strong>To add a friend, use </strong><code>/friend add [username]</code>.<br /><br />`;
+			return buf;
+		}
+		buf += `</h3> `;
+
+		for (const key in categorized) {
+			const friendArray = categorized[key].sort();
+			if (friendArray.length === 0) continue;
+			buf += `<h4>${STATUS_TITLES[key]} (${friendArray.length})</h4>`;
+			for (const friend of friendArray) {
+				const friendID = toID(friend);
+				buf += `<div class="pad"><div>`;
+				buf += this.displayFriend(friendID, loginTimes[friendID]);
+				buf += `</div></div>`;
+			}
+		}
+
+		return buf;
+	}
+	// much more info redacted
+	async visualizePublicList(userid: ID) {
+		const friends: string[] = (await Chat.Friends.getFriends(userid) as any[]).map(f => f.friend);
+		let buf = `<h3>${userid}'s friends:</h3><hr />`;
+		if (!friends.length) {
+			buf += `None.`;
+			return buf;
+		}
+		for (const friend of friends) {
+			buf += `- <username>${friend}</username><br />`;
+		}
+		return buf;
+	}
+	displayFriend(userid: ID, login?: number) {
+		const user = Users.getExact(userid); // we want this to be exact
+		const name = Utils.escapeHTML(user ? user.name : userid);
+		const statusType = user?.connected ?
+			`<strong style="color:${STATUS_COLORS[user.statusType]}">\u25C9 ${STATUS_TITLES[user.statusType]}</strong>` :
+			'\u25CC Offline';
+		let buf = user ?
+			`<span class="username"> <username>${name}</username></span><span><small> (${statusType})</small></span>` :
+			Utils.html`<i>${name}</i> <small>(${statusType})</small>`;
+		buf += `<br />`;
+
+		const curUser = Users.get(userid); // might be an alt
+		if (user) {
+			if (user.userMessage) buf += Utils.html`Status: <i>${user.userMessage}</i><br />`;
+		} else if (curUser && curUser.id !== userid) {
+			buf += `<small>On an alternate account</small><br />`;
+		}
+		if (login && typeof login === 'number' && !user?.connected) {
+			// THIS IS A TERRIBLE HACK BUT IT WORKS OKAY
+			const time = Chat.toTimestamp(new Date(Number(login)), {human: true});
+			buf += `Last login: ${time.split(' ').reverse().join(', on ')}`;
+			buf += ` (${Chat.toDurationString(Date.now() - login, {precision: 1})} ago)`;
+		} else if (typeof login === 'string') {
+			buf += `${login}`;
+		}
+		buf = `<div class="infobox">${buf}</div>`;
+		return toLink(buf);
+	}
+	checkCanUse(context: Chat.CommandContext | Chat.PageContext) {
+		const user = context.user;
+		if (user.locked || user.namelocked || user.semilocked || user.permalocked) {
```

Comment:
Why check `.permalocked`? permalocked users will always be locked anyway.

Reply:
Being extra careful, mostly.

Reply:
If permalocked users don't have `.locked` truthy, there is a much larger issue than them being able to use the friends list.